rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,845 @@
+  requires dedicated support because the potential UB must be reflected in the
+  havoc.
+- **`modifies` clauses over patterns.** Modifies clauses mention values bound in
+  the function header and as a user I would expect that if I use a pattern in
+  the function header then I can use the names bound in that pattern as lvalues
+  in the `modifies` clause. However `modifies` clauses are implemented as
+  `assigns` clauses in CBMC which does not have a notion of function header
+  patterns. Thus it is necessary to project any `modifies` ranges deeper by the
+  fields used in the matched pattern.
+
+<!-- 
+- Is there any part of the design that you expect to resolve through the RFC process?
+- What kind of user feedback do you expect to gather before stabilization? How will this impact your design? 
+-->
+
+## Future possibilities
+
+<!-- For Developers -->
+
+- **Quantifiers:** Quantifiers are like logic-level loops and a powerful
+  reasoning helper. CBMC has support for both `exists` and `forall`, but the
+  code generation is difficult. The most ergonomic and easy way to implement
+  quantifiers on the Rust side is as higher-order functions taking `Fn(T) ->
+  bool`, where `T` is some arbitrary type that can be quantified over. This
+  interface is familiar to developers, but the code generation is tricky, as
+  CBMC level quantifiers only allow certain kinds of expressions. This
+  necessitates a rewrite of the `Fn` closure to a compliant expression.
+- Letting the user supply the **harnesses for checking contracts** is a source of
+  unsoundness, if corner cases are not adequately covered. Ideally Kani would
+  generate the check harness automatically, but this is difficult both because
+  heap data structures are potentially infinite, and also because it must observe
+  user-level type invariants.
+
+  A complete solution for this is not known to us but there are ongoing
+  investigations into harness generation mechanisms in CBMC.
+
+  Environments that are non-inductive could be created from the type as the
+  safe Rust type constraints describe a finite space.
+
+  For dealing with pointers one applicable mechanism could be *memory
+  predicates* to declaratively describe the state of the heap both before and
+  after the function call. 
+  
+  In CBMC's implementation memory predicates are part of the pre/postconditions.
+  This does not easily translate to Kani, since we handle pre/postconditions
+  manually and mainly in proc-macros. There are multiple ways to bridge this
+  gap, perhaps the easiest being to add memory predicates *separately* to Kani
+  instead of as part of pre/postconditions, so they can be handled by forwarding
+  them to CBMC. However this is also tricky, because memory predicates are used
+  to describe pointers and pointers only. Meaning that if they are encapsulated
+  in a structure (such as `Vec` or `RefCell`) there is no way of specifying the
```

Comment:
It's even worse than than, if you have an unbounded vector of raw pointers to lists, you would have to use quantifiers and pointer predicates and we don't know how to handle that just yet